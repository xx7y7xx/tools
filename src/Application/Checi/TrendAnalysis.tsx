import React, { useState, useEffect } from 'react';
import { Card, Space, Input, Typography, Spin, Alert, List } from 'antd';
import { DatabaseOutlined, LineChartOutlined } from '@ant-design/icons';
import { Line } from 'react-chartjs-2';
import {
  Chart as ChartJS,
  CategoryScale,
  LinearScale,
  PointElement,
  LineElement,
  Title as ChartTitle,
  Tooltip,
  Legend,
} from 'chart.js';
import {
  fetchAllHistoricalData,
  extractTrainTrendsFromHistoricalData,
  TrainInfo,
  HistoricalTrainsData,
} from '../../services/trainsData';
import { recentDates } from './config';

ChartJS.register(
  CategoryScale,
  LinearScale,
  PointElement,
  LineElement,
  ChartTitle,
  Tooltip,
  Legend
);

const { Search } = Input;
const { Title, Paragraph } = Typography;

interface TrendAnalysisProps {
  onError: (error: string | null) => void;
  onLoadingChange: (loading: boolean) => void;
}

const TrendAnalysis: React.FC<TrendAnalysisProps> = ({
  onError,
  onLoadingChange,
}) => {
  const [loading, setLoading] = useState(false);
  const [historicalData, setHistoricalData] = useState<HistoricalTrainsData>(
    {}
  );
  const [trendData, setTrendData] = useState<
    { date: string; total_num: number; trainInfo?: TrainInfo }[]
  >([]);
  const [selectedTrain, setSelectedTrain] = useState<string>('');

  const hasHistoricalData =
    historicalData && Object.keys(historicalData).length > 0;

  // Load historical data when component mounts
  useEffect(() => {
    loadHistoricalData();
  }, []);

  const loadHistoricalData = async () => {
    setLoading(true);
    onLoadingChange(true);
    onError(null);

    try {
      console.log('Loading historical data for trend analysis...');
      const data = await fetchAllHistoricalData(recentDates);
      setHistoricalData(data || {});
      console.log('Historical data loaded successfully');
    } catch (err) {
      console.error('Failed to load historical data:', err);
      onError(err instanceof Error ? err.message : '加载历史数据失败');
    } finally {
      setLoading(false);
      onLoadingChange(false);
    }
  };

  const handleSearch = () => {
    if (!selectedTrain.trim()) {
      onError('请输入列车号');
      return;
    }

    if (!hasHistoricalData) {
      onError('历史数据未加载完成，请稍后再试');
      return;
    }

    // Analyze trends for the selected train
    const trends = handleAnalyzeTrainTrends(selectedTrain);
    setTrendData(trends);
  };

  // Analyze trends for a specific train (using preloaded historical data)
  const handleAnalyzeTrainTrends = (trainCode: string) => {
    if (!trainCode.trim()) {
      onError('请输入列车号');
      return [];
    }

    if (!hasHistoricalData) {
      onError('历史数据尚未加载完成，请稍候再试');
      return [];
    }

    const trends = extractTrainTrendsFromHistoricalData(
      trainCode,
      historicalData
    );
    setTrendData(trends);
    setSelectedTrain(trainCode);
    return trends;
  };

  // Prepare chart data
  const chartData = {
    labels: trendData.map((item) => {
      const date = item.date;
      return `${date.slice(4, 6)}/${date.slice(6, 8)}`;
    }),
    datasets: [
      {
        label: `列车 ${selectedTrain} 的 total_num 趋势`,
        data: trendData.map((item) => item.total_num),
        borderColor: 'rgb(75, 192, 192)',
        backgroundColor: 'rgba(75, 192, 192, 0.2)',
        tension: 0.1,
      },
    ],
  };

  const chartOptions = {
    responsive: true,
    plugins: {
      legend: {
        position: 'top' as const,
      },
      title: {
        display: true,
        text: `列车 ${selectedTrain} 数据趋势分析`,
      },
    },
    scales: {
      y: {
        beginAtZero: true,
        title: {
          display: true,
          text: 'Total Number',
        },
      },
      x: {
        title: {
          display: true,
          text: '日期 (MM/DD)',
        },
      },
    },
  };

  return (
    <Space direction="vertical" size="large" style={{ width: '100%' }}>
      {/* Historical Data Status */}
      <Card
        size="small"
        style={{ background: '#f6ffed', border: '1px solid #b7eb8f' }}
      >
        <Space>
          <DatabaseOutlined style={{ color: '#52c41a' }} />
          <span>
            {loading ? (
              <span>正在加载历史数据... ({recentDates.length} 个日期)</span>
            ) : hasHistoricalData ? (
              <span>
                历史数据已加载完成 ({Object.keys(historicalData).length}/
                {recentDates.length} 个日期) - 可进行即时趋势分析
              </span>
            ) : (
              <span>历史数据加载中...</span>
            )}
          </span>
        </Space>
      </Card>

      <div>
        <Title level={4}>列车趋势分析 (基于预加载数据)</Title>
        <Paragraph>
          输入列车号即可查看其在所有历史日期的 total_num 变化趋势 (如: 1461,
          1462)
          <br />
          <strong>数据范围:</strong> 2024年7月至2025年6月 (共
          {recentDates.length}个日期)
          <br />
          <strong>优势:</strong> 所有历史数据已预加载，趋势分析瞬间完成！
        </Paragraph>
        <Search
          placeholder="输入列车号 (如: 1461)"
          allowClear
          enterButton="分析趋势"
          size="large"
          onSearch={handleAnalyzeTrainTrends}
          disabled={!hasHistoricalData}
        />
      </div>

      {/* Trend Chart Display */}
      {trendData.length > 0 && selectedTrain && (
        <Card
          title={`📈 列车 ${selectedTrain} 趋势分析`}
          style={{ marginTop: '20px' }}
        >
          <div style={{ marginBottom: '20px' }}>
            <Paragraph>
              <strong>路线:</strong>{' '}
              {trendData.find((t) => t.trainInfo)?.trainInfo?.from_station}→{' '}
              {trendData.find((t) => t.trainInfo)?.trainInfo?.to_station}
            </Paragraph>
            <Paragraph>
              <strong>数据范围:</strong> {trendData.length} 个日期 (有效数据:{' '}
              {trendData.filter((t) => t.total_num > 0).length} 个)
            </Paragraph>
            <Paragraph>
              <strong>最大值:</strong>{' '}
              {Math.max(...trendData.map((t) => t.total_num))} |{' '}
              <strong>最小值:</strong>{' '}
              {Math.min(...trendData.map((t) => t.total_num))} |{' '}
              <strong>平均值:</strong>{' '}
              {(
                trendData.reduce((sum, t) => sum + t.total_num, 0) /
                trendData.length
              ).toFixed(1)}
            </Paragraph>
          </div>
          <div style={{ height: '400px' }}>
            <Line data={chartData} options={chartOptions} />
          </div>

          {/* Data Table */}
          <div style={{ marginTop: '20px' }}>
            <Title level={5}>详细数据</Title>
            <List
              size="small"
              dataSource={trendData.filter((t) => t.total_num > 0)}
              renderItem={(item) => (
                <List.Item key={item.date}>
                  <span>
                    <strong>{item.date}:</strong> total_num = {item.total_num}
                    {item.trainInfo && (
                      <span style={{ marginLeft: '10px', color: '#666' }}>
                        (车次: {item.trainInfo.train_no})
                      </span>
                    )}
                  </span>
                </List.Item>
              )}
              pagination={{ pageSize: 10, size: 'small' }}
            />
          </div>
        </Card>
      )}
    </Space>
  );
};

export default TrendAnalysis;
