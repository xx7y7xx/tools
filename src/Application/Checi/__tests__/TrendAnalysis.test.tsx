import React from 'react';
import { render, waitFor } from '@testing-library/react';
import TrendAnalysis from '../TrendAnalysis';

jest.mock('axios', () => ({
  get: jest.fn(),
}));

jest.mock('../../../services/trainsData', () => ({
  ...jest.requireActual('../../../services/trainsData'),
  fetchAllHistoricalData: jest.fn(() =>
    Promise.resolve({
      '20241120': {
        '1461': {
          from_station: 'A',
          to_station: 'B',
          station_train_code: '1461',
          total_num: '5',
          train_no: 'T100',
        },
      },
    })
  ),
  extractTrainTrendsFromHistoricalData: jest.fn(() => [
    {
      date: '20241120',
      total_num: 5,
      trainInfo: {
        from_station: 'A',
        to_station: 'B',
        station_train_code: '1461',
        total_num: '5',
        train_no: 'T100',
      },
    },
  ]),
}));

jest.mock('../config', () => ({
  recentDates: ['20241120'],
}));

describe('TrendAnalysis', () => {
  it('matches snapshot', async () => {
    const onError = jest.fn();
    const onLoadingChange = jest.fn();
    const { container } = render(
      <TrendAnalysis onError={onError} onLoadingChange={onLoadingChange} />
    );
    // Wait for useEffect to finish
    await waitFor(() => expect(onLoadingChange).toHaveBeenCalled());
    expect(container).toMatchSnapshot();
  });
});
