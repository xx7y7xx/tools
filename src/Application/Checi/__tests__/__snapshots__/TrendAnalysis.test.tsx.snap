// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrendAnalysis matches snapshot 1`] = `
<div>
  <div
    class="ant-space css-dev-only-do-not-override-1uq9j6g ant-space-vertical ant-space-gap-row-large ant-space-gap-col-large"
    style="width: 100%;"
  >
    <div
      class="ant-space-item"
    >
      <div
        class="ant-card ant-card-bordered ant-card-small css-dev-only-do-not-override-1uq9j6g"
        style="background: rgb(246, 255, 237); border: 1px solid #b7eb8f;"
      >
        <div
          class="ant-card-body"
        >
          <div
            class="ant-space css-dev-only-do-not-override-1uq9j6g ant-space-horizontal ant-space-align-center ant-space-gap-row-small ant-space-gap-col-small"
          >
            <div
              class="ant-space-item"
            >
              <span
                aria-label="database"
                class="anticon anticon-database"
                role="img"
                style="color: rgb(82, 196, 26);"
              >
                <svg
                  aria-hidden="true"
                  data-icon="database"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M832 64H192c-17.7 0-32 14.3-32 32v832c0 17.7 14.3 32 32 32h640c17.7 0 32-14.3 32-32V96c0-17.7-14.3-32-32-32zm-600 72h560v208H232V136zm560 480H232V408h560v208zm0 272H232V680h560v208zM304 240a40 40 0 1080 0 40 40 0 10-80 0zm0 272a40 40 0 1080 0 40 40 0 10-80 0zm0 272a40 40 0 1080 0 40 40 0 10-80 0z"
                  />
                </svg>
              </span>
            </div>
            <div
              class="ant-space-item"
            >
              <span>
                <span>
                  正在加载历史数据... (
                  1
                   个日期)
                </span>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="ant-space-item"
    >
      <div>
        <h4
          class="ant-typography css-dev-only-do-not-override-1uq9j6g"
        >
          列车趋势分析 (基于预加载数据)
        </h4>
        <div
          class="ant-typography css-dev-only-do-not-override-1uq9j6g"
        >
          输入列车号即可查看其在所有历史日期的 total_num 变化趋势 (如: 1461, 1462)
          <br />
          <strong>
            数据范围:
          </strong>
           2024年7月至2025年6月 (共
          1
          个日期)
          <br />
          <strong>
            优势:
          </strong>
           所有历史数据已预加载，趋势分析瞬间完成！
        </div>
        <span
          class="ant-input-group-wrapper ant-input-group-wrapper-disabled ant-input-group-wrapper-lg ant-input-group-wrapper-outlined css-dev-only-do-not-override-1uq9j6g ant-input-search ant-input-search-large ant-input-search-with-button"
        >
          <span
            class="ant-input-wrapper ant-input-group css-dev-only-do-not-override-1uq9j6g"
          >
            <span
              class="ant-input-affix-wrapper ant-input-disabled ant-input-affix-wrapper-disabled ant-input-affix-wrapper-lg css-dev-only-do-not-override-1uq9j6g ant-input-outlined"
            >
              <input
                class="ant-input ant-input-disabled ant-input-lg css-dev-only-do-not-override-1uq9j6g"
                disabled=""
                placeholder="输入列车号 (如: 1461)"
                type="text"
                value=""
              />
              <span
                class="ant-input-suffix"
              >
                <span
                  class="ant-input-clear-icon ant-input-clear-icon-hidden"
                  role="button"
                  tabindex="-1"
                >
                  <span
                    aria-label="close-circle"
                    class="anticon anticon-close-circle"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="close-circle"
                      fill="currentColor"
                      fill-rule="evenodd"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M512 64c247.4 0 448 200.6 448 448S759.4 960 512 960 64 759.4 64 512 264.6 64 512 64zm127.98 274.82h-.04l-.08.06L512 466.75 384.14 338.88c-.04-.05-.06-.06-.08-.06a.12.12 0 00-.07 0c-.03 0-.05.01-.09.05l-45.02 45.02a.2.2 0 00-.05.09.12.12 0 000 .07v.02a.27.27 0 00.06.06L466.75 512 338.88 639.86c-.05.04-.06.06-.06.08a.12.12 0 000 .07c0 .03.01.05.05.09l45.02 45.02a.2.2 0 00.09.05.12.12 0 00.07 0c.02 0 .04-.01.08-.05L512 557.25l127.86 127.87c.04.04.06.05.08.05a.12.12 0 00.07 0c.03 0 .05-.01.09-.05l45.02-45.02a.2.2 0 00.05-.09.12.12 0 000-.07v-.02a.27.27 0 00-.05-.06L557.25 512l127.87-127.86c.04-.04.05-.06.05-.08a.12.12 0 000-.07c0-.03-.01-.05-.05-.09l-45.02-45.02a.2.2 0 00-.09-.05.12.12 0 00-.07 0z"
                      />
                    </svg>
                  </span>
                </span>
              </span>
            </span>
            <span
              class="ant-input-group-addon"
            >
              <button
                class="ant-btn css-dev-only-do-not-override-1uq9j6g ant-btn-primary ant-btn-lg ant-input-search-button"
                disabled=""
                type="button"
              >
                <span>
                  分析趋势
                </span>
              </button>
            </span>
          </span>
        </span>
      </div>
    </div>
  </div>
</div>
`;
